runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java

```diff
@@ -187,11 +209,152 @@ public void flatMap(T t, Collector<T> collector) throws Exception {
+    try {
+      inputCoder = (KvCoder<K, V>) CoderTranslation.fromProto(
+              inputCoderProto,
+              rehydratedComponents);
+    } catch (IOException e) {
+      throw new IllegalStateException(
+              String.format("Unable to hydrate GroupByKey input coder %s.",
+                      inputCoderProto),
+              e);
+    }
+
+    WindowingStrategy<Object, BoundedWindow> windowingStrategy;
+    try {
+      windowingStrategy =
+              (WindowingStrategy<Object, BoundedWindow>) WindowingStrategyTranslation.fromProto(
+                      windowingStrategyProto,
+                      rehydratedComponents);
+    } catch (InvalidProtocolBufferException e) {
+      throw new IllegalStateException(
+              String.format("Unable to hydrate GroupByKey windowing strategy %s.",
+                      windowingStrategyProto),
+              e);
+    }
+
+
+    SingletonKeyedWorkItemCoder<K, V> workItemCoder = SingletonKeyedWorkItemCoder.of(
+            inputCoder.getKeyCoder(),
+            inputCoder.getValueCoder(),
+            windowingStrategy.getWindowFn().windowCoder());
+
+    WindowedValue.
+            FullWindowedValueCoder<SingletonKeyedWorkItem<K, V>> windowedWorkItemCoder =
+            WindowedValue.getFullCoder(
+                    workItemCoder,
+                    windowingStrategy.getWindowFn().windowCoder());
+
+    CoderTypeInformation<WindowedValue<SingletonKeyedWorkItem<K, V>>> workItemTypeInfo =
+            new CoderTypeInformation<>(windowedWorkItemCoder);
+
+    DataStream<WindowedValue<SingletonKeyedWorkItem<K, V>>> workItemStream =
+            inputDataStream
+                    .flatMap(new FlinkStreamingTransformTranslators.ToKeyedWorkItem<>())
+                    .returns(workItemTypeInfo)
+                    .name("ToKeyedWorkItem");
+
+    KeyedStream<WindowedValue<SingletonKeyedWorkItem<K, V>>, ByteBuffer>
+            keyedWorkItemStream =
+            workItemStream.keyBy(new WorkItemKeySelector<>(inputCoder.getKeyCoder()));
+
+    SystemReduceFn<K, V, Iterable<V>, Iterable<V>, BoundedWindow> reduceFn =
+            SystemReduceFn.buffering(inputCoder.getValueCoder());
+
+    //Coder<WindowedValue<KV<K, Iterable<V>>>> outputCoder =
+    //        context.getCoder(context.getOutput(transform));
+    Coder<KV<K, Iterable<V>>> outputCoder;
+    try {
+      outputCoder = (Coder<KV<K, Iterable<V>>>) CoderTranslation.fromProto(
+              outputCoderProto,
+              rehydratedComponents);
+    } catch (IOException e) {
+      throw new IllegalStateException(
+              String.format("Unable to hydrate GroupByKey input coder %s.",
+                      inputCoderProto),
+              e);
+    }
+
+    //TypeInformation<WindowedValue<KV<K, Iterable<V>>>> outputTypeInfo =
+    //        context.getTypeInfo(context.getOutput(transform));
+    // TODO: use output windowing strategy
+    WindowedValue.FullWindowedValueCoder<?> windowedValueOutputCoder =
```

Comment:
Should this use the windowing strategy from the output pcollection instead?

Reply:
This should be correct as the values we get back from an SDK harness are always WindowedValue<T>s.